server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -99,95 +206,219 @@ public DeferringBucketCollector getDeferringCollector() {
+             * Increase the rounding of {@code owningBucketOrd} using
+             * estimated, bucket counts, {@link #rebucket() rebucketing} the all
+             * buckets if the estimated number of wasted buckets is too high.
+             */
+            private byte increaseRounding(long owningBucketOrd, int oldEstimatedBucketCount, long newKey, byte oldRounding) {
+                if (oldRounding >= roundingInfos.length - 1) {
+                    return oldRounding;
+                }
+                if (bounds.size() < owningBucketOrd * 2 + 2) {
+                    long oldSize = bounds.size();
+                    bounds = context.bigArrays().grow(bounds, owningBucketOrd * 2 + 2);
+                    for (long b = oldSize; b < bounds.size(); b++) {
+                        bounds.set(b, (b & 1L) == 0L ? Long.MAX_VALUE : Long.MIN_VALUE);
                     }
-                    bucketOrds = newBucketOrds;
                 }
+                long min = Math.min(bounds.get(owningBucketOrd * 2), newKey);
+                bounds.set(owningBucketOrd * 2, min);
+                long max = Math.max(bounds.get(owningBucketOrd * 2 + 1), newKey);
+                bounds.set(owningBucketOrd * 2 + 1, max);
+                if (oldEstimatedBucketCount <= targetBuckets * roundingInfos[oldRounding].getMaximumInnerInterval()
+                        && max - min <= targetBuckets * roundingInfos[oldRounding].getMaximumRoughEstimateDurationMillis()) {
+                    return oldRounding;
+                }
+                long oldRoughDuration = roundingInfos[oldRounding].roughEstimateDurationMillis;
+                byte newRounding = oldRounding;
+                int newEstimatedBucketCount;
+                do {
+                    newRounding++;
+                    double ratio = (double) oldRoughDuration / (double) roundingInfos[newRounding].getRoughEstimateDurationMillis();
+                    newEstimatedBucketCount = (int) Math.ceil(oldEstimatedBucketCount * ratio);
+                } while (newRounding < roundingInfos.length - 1 && (
+                    newEstimatedBucketCount > targetBuckets * roundingInfos[newRounding].getMaximumInnerInterval()
+                        || max - min > targetBuckets * roundingInfos[newRounding].getMaximumRoughEstimateDurationMillis()));
+                setRounding(owningBucketOrd, newRounding);
+                bounds.set(owningBucketOrd * 2, preparedRoundings[newRounding].round(bounds.get(owningBucketOrd * 2)));
+                bounds.set(owningBucketOrd * 2 + 1, preparedRoundings[newRounding].round(bounds.get(owningBucketOrd * 2 + 1)));
+                wastedBucketsOverestimate += oldEstimatedBucketCount - newEstimatedBucketCount;
+                if (wastedBucketsOverestimate > nextRebucketAt) {
+                    rebucket();
+                    // Bump the threshold for the next rebucketing
+                    wastedBucketsOverestimate = 0;
+                    nextRebucketAt *= 2;
```

Comment:
What's the rationale behind doubling the rebucketing threshold?  Shouldn't it stay constant since the goal of the agg is to produce a relatively constant number of buckets?  As we rebucket and move to coarser intervals, the number of buckets will reduce (and wasted buckets should get cleaned up over time)?

Reply:
My idea was to always produce good buckets in the end but to allow more and more "wasted" buckets as we end up cleaning up more and more. With the default as it stands when the agg is at the top level we don't clean up any of the "wasted" buckets until right before we return the aggs.

This bumping is more to defend against having large numbers of `owningBucketOrd`s that cause us to have to rebucket many many times. To sort of blunt the `O(n^2)` nature of it all. We'd still to the `O(n)` rebucketing many times. Just fewer times as you accumulate more and more buckets.

I did think about making it a constant budget. And that works pretty well too because we naturally rebucket less and less as we promote rounding because the buckets are bigger. But if we have a ton of `owningBucketOrds` I'm kind of worried.